src/common/dsp/effect/chowdsp/exciter/Exciter.cpp

```diff
@@ -0,0 +1,173 @@
+/*
+** Surge Synthesizer is Free and Open Source Software
+**
+** Surge is made available under the Gnu General Public License, v3.0
+** https://www.gnu.org/licenses/gpl-3.0.en.html
+**
+** Copyright 2004-2020 by various individuals as described by the Git transaction log
+**
+** All source at: https://github.com/surge-synthesizer/surge.git
+**
+** Surge was a commercial product from 2004-2018, with Copyright and ownership
+** in that period held by Claes Johanson at Vember Audio. Claes made Surge
+** open source in September 2018.
+*/
+
+#include "Exciter.h"
+
+namespace
+{
+constexpr double low_freq = 500.0;
+constexpr double high_freq = 10000.0;
+constexpr double q_val = 0.7071;
+} // namespace
+
+namespace chowdsp
+{
+
+Exciter::Exciter(SurgeStorage *storage, FxStorage *fxdata, pdata *pd) : Effect(storage, fxdata, pd)
+{
+    wet_gain.set_blocksize(BLOCK_SIZE);
+    drive_gain.set_blocksize(BLOCK_SIZE);
+}
+
+Exciter::~Exciter() {}
+
+void Exciter::init()
+{
+    toneFilter.suspend();
+    toneFilter.coeff_HP(M_PI, q_val);
+    toneFilter.coeff_instantize();
+
+    levelDetector.reset(samplerate);
+
+    drive_gain.set_target(1.0f);
+    wet_gain.set_target(0.0f);
+}
+
+void Exciter::process(float *dataL, float *dataR)
+{
+    set_params();
+
+    // copy dry signal
+    copy_block(dataL, dryL, BLOCK_SIZE_QUAD);
+    copy_block(dataR, dryR, BLOCK_SIZE_QUAD);
+
+    drive_gain.multiply_2_blocks(dataL, dataR, BLOCK_SIZE_QUAD * os.getOSRatio());
+    os.upsample(dataL, dataR);
+
+    for (int k = 0; k < os.getUpBlockSize(); k++)
+        process_sample(os.leftUp[k], os.rightUp[k]);
+
+    os.downsample(dataL, dataR);
+
+    // dry/wet process
+    wet_gain.multiply_2_blocks(dataL, dataR, BLOCK_SIZE_QUAD);
+    add_block(dataL, dryL, dataL, BLOCK_SIZE_QUAD);
+    add_block(dataR, dryR, dataR, BLOCK_SIZE_QUAD);
+}
+
+void Exciter::set_params()
+{
+    // "Tone" param
+    auto cutoff = low_freq * std::pow(high_freq / low_freq, *f[exciter_tone]);
+    cutoff = limit_range(cutoff, 10.0, samplerate * 0.48);
+    auto omega_factor = samplerate_inv * 2.0 * M_PI / (double)os.getOSRatio();
+    toneFilter.coeff_HP(cutoff * omega_factor, q_val);
+
+    // "Drive" param
+    auto drive_makeup = std::pow(0.2f, 1.f - *f[exciter_tone]);
+    auto drive = 8.f * std::pow(*f[exciter_drive], 1.5f) * drive_makeup;
+    drive_gain.set_target_smoothed(drive);
+
+    // attack/release params
+    auto attack_ms = std::pow(2.0f, fxdata->p[exciter_att].displayInfo.b * *f[exciter_att]);
+    auto release_ms = 10.0f * std::pow(2.0f, fxdata->p[exciter_rel].displayInfo.b * *f[exciter_rel]);
```

Comment:
I was able to set up the parameter types, and they seem to be working as expected. I'm using the `ATwoBx` scaling type. I was wondering if there's an easier to convert to/from the normalised value to the scaled value? This is my current hack, but it should be generalisable to any exponential parameter, like:
```
denormalised_value = param->displayInfo.a * std::pow(2.0f, param->displayInfo.b * normalised_value);
```

Reply:
I think you could reuse `ct_reverbpredelaytime`, which gives you 0 to 2 seconds of range that is exponentially scaled. Or you could just create your own params based on that one.

Reply:
Oh ha we added the ATwoBx for a formatting only thing long after the params. I never ever even thought about using the displayInfo values in the DSP code. 

There’s nothing wrong with it of course - in fact it is rather clever and we could easily add it to param (like “getDisplayNormalizedFloat” which returns the scaled float if you are one of the display types exactly as you suggest) and that’s a very good idea. But like I said it isn’t one I ever thought about at all!

Context: All the displayInfo stuff was retrofitted onto param to remove essentially a bunch of duplicated switch statements and allow us to do the reverse string -> param for typein. I always thought of it purely in gui / daw land. This is why having people who are new to the code read it is so valuable.

I propose we merge this - your change is fine - so I’ll press the button.  But if you agree with assessment above add another issue to add that function point and then we can retrofit.